### ISSUE TITLE
Visual prompt deep (vpt_emb_deep) same for every sample?

### ISSUE THREAD

**Issue by tsly123** [sIncerass/MVLPT#4]
Hi @bronyayang, @sIncerass 

I try to extract the prompt tokens (in VPT) for every sample in evaluation mode with pre-trained prompt_learner, e.g., `--eval-only` and `--model-dir`

The `vpt_emb_deep` tokens [https://github.com/sIncerass/MVLPT/blob/main/trainers/mvlpt.py#L80](https://github.com/sIncerass/MVLPT/blob/main/trainers/mvlpt.py#L80) are same for every sample.

Do they suppose to be the same? Or did I do something wrong? See `######### EXTRACT FEATURE FOR EACH LAYER #########` below.

Thank you.

```
class ImageEncoder(nn.Module):
    def __init__(self, clip_model, mvlpt_model):
        super().__init__()
        # HACK: Assume all is vision transformer
        self.visual = clip_model.visual
        self.mvlpt_model = mvlpt_model

    def forward(self, x: torch.Tensor, vpt_embeddings=None, vpt_embeddings_deep=None, extract_layer=True):
        x = self.visual.conv1(x)  # shape = [*, width, grid, grid]
        x = x.reshape(x.shape[0], x.shape[1], -1)  # shape = [*, width, grid ** 2]
        x = x.permute(0, 2, 1)  # shape = [*, grid ** 2, width]
        x = torch.cat([self.visual.class_embedding.to(x.dtype) + torch.zeros(x.shape[0], 1, x.shape[-1], dtype=x.dtype, device=x.device), x], dim=1)  # shape = [*, grid ** 2 + 1, width]
        x = x + self.visual.positional_embedding.to(x.dtype)
        x = self.visual.ln_pre(x)

        B = x.shape[0]

        x, ctx = self.mvlpt_model.forward_vpt(x, vpt_embeddings)
        # ext_ctx = ctx.detach().cpu().numpy()
        x = x.permute(1, 0, 2)  # NLD -> LND

        ext_x = []
        ext_vpt_emb_deep = []
        ext_visual = []

        if self.mvlpt_model.vpt_deep and (vpt_embeddings_deep is not None or self.mvlpt_model.vpt_embeddings_deep is not None):
            if vpt_embeddings_deep is None:
                vpt_embeddings_deep = self.mvlpt_model.vpt_embeddings_deep

            for layer_idx in range(self.visual.transformer.layers):
                layer = self.visual.transformer.resblocks[layer_idx]
                
                if layer_idx == 0:
                    x = layer(x)
                elif layer_idx <= vpt_embeddings_deep.shape[0]:
                    vpt_emb_deep = self.mvlpt_model.vpt_dropout(self.mvlpt_model.vpt_proj(
                        vpt_embeddings_deep[layer_idx-1]).expand(B, -1, -1)).to(x.dtype)

                    vpt_emb_deep = vpt_emb_deep.permute(1, 0, 2)  # NLD -> LND
                    x = torch.cat((
                        x[:1, :, :],
                        vpt_emb_deep,
                        x[(1+self.mvlpt_model.vpt_n_ctx):, :, :]
                    ), dim=0)
                    x = layer(x)

                    ######### EXTRACT FEATURE FOR EACH LAYER #########
                    if extract_layer:
                        '''
                        extract features
                        ext_x: layer's output feature, [L, N, D]
                        ext_vpt_emb_deep: input prompt vector, [L, N, D]
                        ext_visual: input image patch, [L, N, D]
                        '''
                        ext_x.append(x.detach().permute(1, 0, 2).unsqueeze(0))
                        ext_vpt_emb_deep.append(vpt_emb_deep.detach().permute(1, 0, 2).unsqueeze(0))
                        ext_visual.append(x[(1+self.mvlpt_model.vpt_n_ctx):, :, :].detach().permute(1, 0, 2).unsqueeze(0))

                        # print('ImageEncoder ext_x', self.ext_x.shape)
                        # print('ImageEncoder ext_vpt_emb_deep', self.ext_vpt_emb_deep.shape)
                        # print('ImageEncoder ext_visual', self.ext_visual.shape)
                    ######### END EXTRACT FEATURE FOR EACH LAYER #########
        else:
            x = self.visual.transformer(x)
        x = x.permute(1, 0, 2)  # LND -> NLD

        x = self.visual.ln_post(x[:, 0, :])

        if self.visual.proj is not None:
            x = x @ self.visual.proj

        ######### EXTRACT FEATURE FOR EACH LAYER #########
        if extract_layer:
            ext_x = torch.cat(ext_x, 0)     # [n_layer, B, token_len, dim]
            ext_vpt_emb_deep = torch.cat(ext_vpt_emb_deep, 0)   # [n_layer, B, token_len, dim]
            ext_visual = torch.cat(ext_visual, 0)   # [n_layer, B, token_len, dim]

            # compare prompt input for each layer
            eq = []
            for layer_idx in range(len(ext_vpt_emb_deep)):
                for i in range(len(ext_vpt_emb_deep[layer_idx])):
                    eq.append(torch.equal(ext_vpt_emb_deep[layer_idx][0],
                                          ext_vpt_emb_deep[layer_idx][i]))

            print('ImageEncoder test eq:', eq)  # ALL TRUE for EACH LAYER. WHY???

            return x, ext_x, ext_vpt_emb_deep, ext_visual
        ######### END EXTRACT FEATURE FOR EACH LAYER #########

        else:
            return x
```
